File: src/Entity/Commande.php
<?php

namespace App\Entity;

use App\Repository\CommandeRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

#[ORM\Entity(repositoryClass: CommandeRepository::class)]
#[ORM\Table(name: 'commande')]
class Commande
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'id_commande')]
    private ?int $id = null;

    #[ORM\Column(name: 'date_commande', type: 'datetime_immutable')]
    private \DateTimeImmutable $dateCommande;

    #[ORM\Column(length: 50, options: ['default' => 'en_attente'])]
    private string $statut = 'en_attente';

    #[ORM\Column(type: 'decimal', precision: 10, scale: 2, options: ['default' => 0])]
    private string $total = '0.00';

    #[ORM\Column(name: 'mode_paiement', length: 50)]
    private string $modePaiement = '';

    #[ORM\Column(name: 'adresse_livraison', length: 255)]
    private string $adresseLivraison = '';

    #[ORM\Column(length: 20)]
    private string $telephone = '';

    #[ORM\Column(length: 255)]
    private string $nom = '';

    #[ORM\Column(length: 255)]
    private string $email = '';

    #[ORM\Column(type: 'text', nullable: true)]
    private ?string $message = null;

    // Temporary solution: store product IDs as JSON string
    #[ORM\Column(type: 'text', nullable: true)]
    private ?string $produitsIds = null;

    #[ORM\Column(name: 'coupon_code', length: 64, nullable: true)]
    private ?string $couponCode = null;

    #[ORM\Column(name: 'coupon_discount', type: 'decimal', precision: 10, scale: 2, options: ['default' => 0])]
    private string $couponDiscount = '0.00';

    #[ORM\Column(name: 'estimated_delivery_date', nullable: true)]
    private ?\DateTimeImmutable $estimatedDeliveryDate = null;

    #[ORM\Column(name: 'fraud_score', options: ['default' => 0])]
    private int $fraudScore = 0;

    #[ORM\Column(name: 'base_shipping_cost', type: 'decimal', precision: 10, scale: 2, options: ['default' => 0])]
    private string $baseShippingCost = '0.00';

    #[ORM\ManyToOne(targetEntity: Utilisateur::class)]
    #[ORM\JoinColumn(name: 'id_utilisateur_id', referencedColumnName: 'id_utilisateur', nullable: true, onDelete: 'SET NULL')]
    private ?Utilisateur $utilisateur = null;

    #[ORM\OneToMany(mappedBy: 'commande', targetEntity: OrderShipment::class, cascade: ['persist', 'remove'], orphanRemoval: true)]
    private Collection $shipments;

    public function __construct()
    {
        $this->dateCommande = new \DateTimeImmutable();
        $this->statut = 'en_attente';
        $this->total = '0.00';
        $this->couponDiscount = '0.00';
        $this->fraudScore = 0;
        $this->baseShippingCost = '0.00';
        $this->shipments = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getDateCommande(): \DateTimeImmutable
    {
        return $this->dateCommande;
    }

    protected function setDateCommande(\DateTimeImmutable $dateCommande): static
    {
        $this->dateCommande = $dateCommande;
        return $this;
    }

    public function getStatut(): string
    {
        return $this->statut;
    }

    public function setStatut(string $statut): static
    {
        $this->statut = $statut;
        return $this;
    }

    public function getTotal(): float
    {
        return (float) $this->total;
    }

    public function setTotal(float $total): static
    {
        $this->total = (string) $total;
        return $this;
    }

    public function getModePaiement(): string
    {
        return $this->modePaiement;
    }

    public function setModePaiement(string $modePaiement): static
    {
        $this->modePaiement = $modePaiement;
        return $this;
    }

    public function getAdresseLivraison(): string
    {
        return $this->adresseLivraison;
    }

    public function setAdresseLivraison(string $adresseLivraison): static
    {
        $this->adresseLivraison = $adresseLivraison;
        return $this;
    }

    public function getTelephone(): string
    {
        return $this->telephone;
    }

    public function setTelephone(string $telephone): static
    {
        $this->telephone = $telephone;
        return $this;
    }

    public function getNom(): string
    {
        return $this->nom;
    }

    public function setNom(string $nom): static
    {
        $this->nom = $nom;
        return $this;
    }

    public function getEmail(): string
    {
        return $this->email;
    }

    public function setEmail(string $email): static
    {
        $this->email = $email;
        return $this;
    }

    public function getMessage(): ?string
    {
        return $this->message;
    }

    public function setMessage(?string $message): static
    {
        $this->message = $message;
        return $this;
    }

    public function getProduitsIds(): ?string
    {
        return $this->produitsIds;
    }

    public function setProduitsIds(?string $produitsIds): static
    {
        $this->produitsIds = $produitsIds;
        return $this;
    }

    /**
     * Get products array from JSON string
     */
    public function getProduitsArray(): array
    {
        return $this->produitsIds ? json_decode($this->produitsIds, true) : [];
    }

    /**
     * Set products from array
     */
    public function setProduitsFromArray(array $produits): static
    {
        $this->produitsIds = json_encode(array_values($produits));
        return $this;
    }

    /**
     * Calculate total based on selected products (requires repository)
     */
    public function calculateTotalFromProducts(array $products): float
    {
        $total = 0.0;
        foreach ($products as $product) {
            $total += $product->getPrix();
        }
        return $total;
    }

    public function getCouponCode(): ?string
    {
        return $this->couponCode;
    }

    public function setCouponCode(?string $couponCode): static
    {
        $this->couponCode = $couponCode;
        return $this;
    }

    public function getCouponDiscount(): float
    {
        return (float) ($this->couponDiscount ?? 0.0);
    }

    public function setCouponDiscount(float $couponDiscount): static
    {
        $this->couponDiscount = max(0.0, $couponDiscount);
        return $this;
    }

    public function getEstimatedDeliveryDate(): ?\DateTimeImmutable
    {
        return $this->estimatedDeliveryDate;
    }

    protected function setEstimatedDeliveryDate(?\DateTimeImmutable $estimatedDeliveryDate): static
    {
        $this->estimatedDeliveryDate = $estimatedDeliveryDate;
        return $this;
    }

    public function getFraudScore(): int
    {
        return $this->fraudScore;
    }

    public function setFraudScore(int $fraudScore): static
    {
        $this->fraudScore = max(0, min(100, $fraudScore));
        return $this;
    }

    public function getBaseShippingCost(): float
    {
        return (float) ($this->baseShippingCost ?? 0.0);
    }

    public function setBaseShippingCost(float $baseShippingCost): static
    {
        $this->baseShippingCost = max(0.0, $baseShippingCost);
        return $this;
    }

    public function getUtilisateur(): ?Utilisateur
    {
        return $this->utilisateur;
    }

    public function setUtilisateur(?Utilisateur $utilisateur): static
    {
        $this->utilisateur = $utilisateur;
        return $this;
    }

    /**
     * @return Collection<int, OrderShipment>
     */
    public function getShipments(): Collection
    {
        return $this->shipments;
    }

    public function addShipment(OrderShipment $shipment): static
    {
        if (!$this->shipments->contains($shipment)) {
            $this->shipments->add($shipment);
            $shipment->setCommande($this);
        }
        return $this;
    }

    public function removeShipment(OrderShipment $shipment): static
    {
        if ($this->shipments->removeElement($shipment)) {
            if ($shipment->getCommande() === $this) {
                $shipment->setCommande(null);
            }
        }
        return $this;
    }
}

File: src/main/java/org/example/model/Commande.java
package org.example.model;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;

public class Commande {
    private int id;
    private Client client;
    private List<Produit> produits;
    private LocalDate dateCommande;
    /** Legacy TTC field; kept for older code paths */
    private double total;
    private String statut;

    private Integer utilisateurId;
    private String email;
    private String nom;
    private String telephone;
    private String modePaiement;
    private String adresseLivraison;
    private String message;
    private String produitsIds;
    private String couponCode;
    private double couponDiscount;
    private int fraudScore;
    private double baseShippingCost;
    private LocalDateTime estimatedDeliveryDate;
    private LocalDateTime dateCommandeDateTime;
    private double totalHt;
    private double totalTtc;

    public Commande() {}

    public Commande(int id, Client client, List<Produit> produits, LocalDate dateCommande, double total, String statut) {
        this.id = id;
        this.client = client;
        this.produits = produits;
        this.dateCommande = dateCommande;
        this.total = total;
        this.statut = statut;
        this.totalTtc = total;
    }

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public Client getClient() { return client; }
    public void setClient(Client client) { this.client = client; }

    public List<Produit> getProduits() { return produits; }
    public void setProduits(List<Produit> produits) { this.produits = produits; }

    public LocalDate getDateCommande() { return dateCommande; }
    public void setDateCommande(LocalDate dateCommande) { this.dateCommande = dateCommande; }

    /**
     * Montant TTC affiché aux écrans legacy (tableaux, mails, scoring).
     * Préfère {@link #totalTtc} lorsqu'il a été renseigné par le flux e-commerce.
     */
    public double getTotal() {
        if (totalTtc > 0) {
            return totalTtc;
        }
        return total;
    }

    public void setTotal(double total) {
        this.total = total;
        if (totalTtc <= 0) {
            this.totalTtc = total;
        }
    }

    public String getStatut() { return statut; }
    public void setStatut(String statut) { this.statut = statut; }

    public Integer getUtilisateurId() { return utilisateurId; }
    public void setUtilisateurId(Integer utilisateurId) { this.utilisateurId = utilisateurId; }

    public String getEmail() {
        if (email != null && !email.isBlank()) {
            return email;
        }
        return client != null ? client.getEmail() : null;
    }

    public void setEmail(String email) { this.email = email; }

    public String getNom() {
        if (nom != null && !nom.isBlank()) {
            return nom;
        }
        if (client != null) {
            String n = client.getNom() == null ? "" : client.getNom();
            String p = client.getPrenom() == null ? "" : client.getPrenom();
            return (n + " " + p).trim();
        }
        return "";
    }

    public void setNom(String nom) { this.nom = nom; }

    /** Nom affiché dans les tableaux admin (PropertyValueFactory "clientNom") */
    public String getClientNom() {
        return getNom();
    }

    public String getTelephone() { return telephone; }
    public void setTelephone(String telephone) { this.telephone = telephone; }

    public String getModePaiement() { return modePaiement; }
    public void setModePaiement(String modePaiement) { this.modePaiement = modePaiement; }

    public String getAdresseLivraison() { return adresseLivraison; }
    public void setAdresseLivraison(String adresseLivraison) { this.adresseLivraison = adresseLivraison; }

    public String getMessage() { return message; }
    public void setMessage(String message) { this.message = message; }

    public String getProduitsIds() { return produitsIds; }
    public void setProduitsIds(String produitsIds) { this.produitsIds = produitsIds; }

    public String getCouponCode() { return couponCode; }
    public void setCouponCode(String couponCode) { this.couponCode = couponCode; }

    public double getCouponDiscount() { return couponDiscount; }
    public void setCouponDiscount(double couponDiscount) { this.couponDiscount = couponDiscount; }

    public int getFraudScore() { return fraudScore; }
    public void setFraudScore(int fraudScore) { this.fraudScore = fraudScore; }

    public double getBaseShippingCost() { return baseShippingCost; }
    public void setBaseShippingCost(double baseShippingCost) { this.baseShippingCost = baseShippingCost; }

    public LocalDateTime getEstimatedDeliveryDate() { return estimatedDeliveryDate; }
    public void setEstimatedDeliveryDate(LocalDateTime estimatedDeliveryDate) {
        this.estimatedDeliveryDate = estimatedDeliveryDate;
    }

    public LocalDateTime getDateCommandeDateTime() {
        if (dateCommandeDateTime != null) {
            return dateCommandeDateTime;
        }
        if (dateCommande != null) {
            return dateCommande.atStartOfDay();
        }
        return null;
    }

    public void setDateCommandeDateTime(LocalDateTime dateCommandeDateTime) {
        this.dateCommandeDateTime = dateCommandeDateTime;
    }

    public double getTotalHt() { return totalHt; }

    public void setTotalHt(double totalHt) { this.totalHt = totalHt; }

    public double getTotalTtc() {
        if (totalTtc > 0) {
            return totalTtc;
        }
        return total;
    }

    public void setTotalTtc(double totalTtc) {
        this.totalTtc = totalTtc;
        this.total = totalTtc;
    }

    @Override
    public String toString() {
        if (client != null) {
            return "Commande " + id + " - " + client.getNom() + " " + client.getPrenom();
        }
        return "Commande " + id + " - " + getNom();
    }
}
